### PR TITLE
[SPARK-48503][SQL] Fix invalid scalar subqueries with group-by on non-equivalent columns that were incorrectly allowed

### PR DESCRIPTION
**Pull request by jchen5**
### What changes were proposed in this pull request?

Fixes CheckAnalysis to reject invalid scalar subquery group-bys that were previously allowed and returned wrong results.

For example, this query is not legal and should give an error, but instead we incorrectly allowed it and it returns wrong results prior to this PR (full repro with table data in the jira):

```
select *, (select count(*) from y where y1 > x1 group by y1) from x;
```

It returns two rows, even though there's only one row of x. The correct result is an error, because there is more than one row returned by the scalar subquery.

Another problem case is if the correlation condition is an equality but it's under another operator like an OUTER JOIN or UNION. Various other expressions that are not equi-joins between the inner and outer fields hit this too, e.g. `where y1 + y2 = x1 group by y1`. See the comments in the code and the tests for more examples.

This PR fixes the logic which checks for valid vs invalid group-bys. However, note that this new logic may block some queries that are actually valid, for example `a + 1 = outer(b)` is valid but would be rejected. Therefore, we add a conf flag that can be used to restore the legacy behavior, as well as logging for when the legacy behavior is used and differs from the new behavior. (In general, to accurately run valid queries and reject invalid queries, the check must be moved from compile-time to run-time - see https://issues.apache.org/jira/browse/SPARK-48501.)

This is a longstanding bug. The bug is in CheckAnalysis in checkAggregateInScalarSubquery. It allows grouping columns that are present in correlation predicates, but doesn’t check whether those predicates are equalities -  because when that code was written, non-equality correlation wasn’t allowed. Therefore, this bug has existed since non-equality correlation was added (~2 years ago).

### Why are the changes needed?
Fix invalid queries returning wrong results

### Does this PR introduce _any_ user-facing change?
Yes, block subqueries with invalid group-bys.

### How was this patch tested?
Add tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 1]
```diff
@@ -249,6 +250,73 @@ object SubExprUtils extends PredicateHelper {
...
+   *     (select * from z where z1 = x1) sub on y2 = z2 group by z1) from x;
+   * - An equality under UNION e.g.
+   *   select *, (select count(*) from
+   *     (select * from y where y1 = x1 union all select * from y) group by y1) from x;
+   */
+  def getCorrelatedEquivalentInnerColumns(plan: LogicalPlan): AttributeSet = {
+    plan match {
+      case Filter(cond, child) =>
+        val correlated = AttributeSet(splitConjunctivePredicates(cond)
+          .filter(containsOuter) // TODO: can remove this line to allow e.g. where x = 1 group by x
```
Comment (jchen5) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:277]: I intend to enable that in a separate PR, to reduce risk here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 2]
```diff
@@ -249,6 +250,73 @@ object SubExprUtils extends PredicateHelper {
...
+        _: Sample |
+        _: Sort |
+        _: Window |
+        _: Tail |
+        _: WithCTE |
+        _: Range |
+        _: SubqueryAlias =>
+        AttributeSet(plan.children.flatMap(child => getCorrelatedEquivalentInnerColumns(child)))
+
+      case _ => AttributeSet.empty
```
Comment (jchen5) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:317]: The list of operators handled here is by no means comprehensive and ensuring it covers enough is tricky. I used the list in LogicalPlanVisitor as a starting point, but in my testing I discovered that e.g. SubqueryAlias also needs to be handled to cover cases with FROM subqueries inside the scalar subquery.

Suggestions on other important operators to handle or other potential approaches welcome. (In the long run I think we need to replace this entire check with a runtime check as described in https://issues.apache.org/jira/browse/SPARK-48501, but that's highly nontrivial)

**Comment by jchen5** [comment]
@agubichev @andylam-db @cloud-fan

**Review by agubichev — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**